pkg/track/events.go

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package track
+
+import (
+	"context"
+	"sync"
+
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	"github.com/microsoft/retina/pkg/enricher"
+	"github.com/microsoft/retina/pkg/log"
+)
+
+type Track struct {
+	ctx             context.Context
+	cancel          context.CancelFunc
+	externalChannel chan *v1.Event
+	l               *log.ZapLogger
+	eh              enricher.EnricherInterface
+}
+
+const (
+	DefaultExternalChannelSize = 1000
+)
+
+var (
+	once sync.Once
+	t    *Track
+)
+
+func New() *Track {
+	once.Do(func() {
+		var e *enricher.Enricher
+		if e = enricher.Instance(); e == nil {
```

Comment:
will cause the old error when we tried to compare a nil interface with nil (this will never be true)

Reply:
Instance returns struct pointer, not interface - https://github.com/microsoft/retina/blob/f93df8d2ed54ca8fb77bf40d3dbb314c6275aea0/pkg/enricher/enricher.go#L61

Reply:
Also, enricher should absolutely be a singleton for the entire lifecycle of Retina. It's a key piece of the control plane that needs the ringbuffer to be set and all events to be funneled to the same. The enricher also serves the enriched data to endpoint as metrics. All of this is predicated on a single enricher instance.